### DISCUSSION TITLE
logout should remove auth session data rather than flushing

### DISCUSSION THREAD

**Discussion by ryanmcgrath** [maxcountryman/axum-login#207] [Q&A] [answered]
- [+] I have looked for existing issues (including closed) about this

### Version

0.15

### Platform

All platforms

### Description

Currently, when one signs in with axum-login, session data for auth specifically is written into the session with `self.data_key`. However, when calling `logout`, it's calling `tower-sessions::Session` `flush`, which obliterates the entire session. This was unexpected behavior - the docs don't note that it's destructive with respect to the entire session - and there are scenarios where one might want to have a session record for non-authenticated users while still having extra data for authenticated users.

https://github.com/maxcountryman/axum-login/blob/e9597526afbe6325826b24c04af645f1e52eeae6/axum-login/src/session.rs#L121

https://github.com/maxcountryman/axum-login/blob/e9597526afbe6325826b24c04af645f1e52eeae6/axum-login/src/session.rs#L140-L142

https://github.com/maxcountryman/axum-login/blob/e9597526afbe6325826b24c04af645f1e52eeae6/axum-login/src/session.rs#L135

Is there any design decision/reason that `logout` can't just pop the data held at `self.data_key` (and maybe just provide a convenience method akin to `logout_and_flush` or something)?

**Comment by maxcountryman** [comment, marked as answer]
Please see [the Django reference](https://github.com/django/django/blob/ae10146793dca0e0594c7acdee20ca3810983f39/django/contrib/auth/__init__.py#L171) for the implementation.

This is mentioned in [the Django documentation](https://docs.djangoproject.com/en/5.0/topics/auth/default/#how-to-log-a-user-out), so I agree it would be a good idea to extend the documentation similarly.

**Reply by ryanmcgrath** [reply]
Hmm, good point - I probably did the "add after flushing" routine myself in a past life. Doc improvements would def help for the cases where someone runs into this need.
